Fixing some issues with BufferedMutator

sduskis opened the pull request:
Adding a separate executor for the heapSizeManager synchronized work.
Better accounting for exceptions in issueRequest, such as validation issues.

sduskis commented:
I addressed the globalException issues.  I moved close (although it probably doesn't make a difference).  I wasn't sure what to do about Futures.immediateFailedFuture() as per the comments.

PTAL.

mgarolera commented on bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableBufferedMutator.java:
```diff
@@ -196,13 +209,8 @@ public BigtableBufferedMutator(
-    } catch (IOException e) {
-      LOG.warn("Could not get the host used for writes", options);
-    }
     this.host = host == null ? null : host.toString();
```
Can host be null? Could this be simplified?

mgarolera commented on bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableBufferedMutator.java:
```diff
@@ -147,6 +156,10 @@ public synchronized void operationComplete(long operationSequenceId) {
       return !pendingOperationsWithSize.isEmpty();
     }
+
+    public synchronized long getHeapSize() {
```
Should this be @VisibleForTesting?

mgarolera commented on bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableBufferedMutator.java:
```diff
@@ -347,14 +356,37 @@ private void doMutation(final Mutation mutation) throws RetriesExhaustedWithDeta
+    try {
+      request = batchExecutor.issueRequest(mutation);
+    } catch (RuntimeException e) {
+      // issueReqeust(mutation) could throw an Exception for validation issues. Remove the heapsize
```
issueRequest

mgarolera commented on bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableBufferedMutator.java:
```diff
@@ -347,14 +356,37 @@ private void doMutation(final Mutation mutation) throws RetriesExhaustedWithDeta
-    Futures.addCallback(batchExecutor.issueRequest(mutation), callback);
+    ListenableFuture<? extends GeneratedMessage> request = null;
+    try {
+      request = batchExecutor.issueRequest(mutation);
```
Is this raising an UnsupportedOperationException? (that is the one I found reading). It is a pity that it is an unchecked exception, otherwise it would have been easier to catch this. :/ Maybe it should be added to issueRequest and OperationAdapter.adapt?

sduskis replied:
putAdapter throws IllegalArgumentException.  mutationAdapter and DeleteAdapters throw UnsupportedOperationException.  Perhaps we should discuss checked exceptions for the Adapter framework, but that sounds like something for another time.

AngusDavis replied:
Further, all gRPC errors (e.g., channel closed) are RuntimeException derived.

sduskis commented:
@mgarolera: I fixed the typo and added a bunch of @VisibleForTesting.  PTAL

sduskis commented:
@AngusDavis: Thanks for the pointer on immediateFailFuture.  I updated the code to use it, and update the test to account for the asynchronous processing of the exception (since the handling happens on a different thread)

AngusDavis commented on bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableBufferedMutator.java:
```diff
@@ -347,14 +356,34 @@ private void doMutation(final Mutation mutation) throws RetriesExhaustedWithDeta
+        // Add the exception to the list of global exceptions and immediately throw a
+        // RetriesExhaustedWithDetailsException.
+        addGlobalException(mutation, e);
+        handleExceptions();
```
This now handles the case of concurrent writers, but couldn't a client still specify an error handler other than the default 'throw whenever floats your boat and on a random thread'? I'm not a fan of that error handler so I imagine I would be one of those clients :)

sduskis commented:
@AngusDavis: Good point on the custom exception handler.  I updated the handling in doMutate() - I now return in the catch clause.  I also updated the exception comment.

AngusDavis commented:
LGTM. For a future change, if you make ExecutorService a constructor parameter then you can pass in a directExecutor and remove some sleep calls from tests. Or a FakeExecutorService-sort of class that allows you to execute 'ticks' to verify that it's done async properly.
